File: RoundCorner/RoundCorner/ViewController.m
//
//  ViewController.m
//  RoundCorner
//
//  Created by virusbee on 2021/4/23.
//

#import "ViewController.h"
#import "CALayer+RoundCorner.h"
#import "RoundSectionTableViewController.h"

@interface ViewController ()
@property (nonatomic, strong) UIView *contentView;
@property (nonatomic, strong) UILabel *label;
@property (nonatomic, strong) UIButton *button;
@property (nonatomic, strong) UITextField *textField;
@property (nonatomic, strong) UIView *emptyView;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self.view addSubview:self.contentView];
    [self.contentView addSubview:self.label];
    [self.contentView addSubview:self.button];
    [self.contentView addSubview:self.textField];
    [self.contentView addSubview:self.emptyView];
}

- (UIView *)contentView {
    if (!_contentView) {
        _contentView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 300, 300)];
        _contentView.center = self.view.center;
        RCConfiguration *config = [[RCConfiguration alloc] init];
        config.cornerMask = RCCornerMaskAll;
        config.rect = _contentView.bounds;
        config.cornerRadius = 20;
        config.strokeColor = UIColor.redColor.CGColor;
        [_contentView.layer setRoundCorner:config];
    }
    return _contentView;
}

- (UILabel *)label {
    if (!_label) {
        _label = [[UILabel alloc] initWithFrame:CGRectMake(20, 20, 125, 125)];
        RCConfiguration *config = [[RCConfiguration alloc] init];
        config.cornerMask = RCCornerMaskAll ^ RCCornerMaskMaxXMaxY;
        config.rect = _label.bounds;
        config.cornerRadius = 20;
        config.fillColor = UIColor.redColor.CGColor;
        [_label.layer setRoundCorner:config];
    }
    return _label;
}

- (UIButton *)button {
    if (!_button) {
        _button = [[UIButton alloc] initWithFrame:CGRectMake(155, 20, 125, 125)];
        [_button setTitle:@"Touch Me" forState:UIControlStateNormal];
        [_button addTarget:self action:@selector(showRoundSectionTableViewController) forControlEvents:UIControlEventTouchUpInside];
        RCConfiguration *config = [[RCConfiguration alloc] init];
        config.cornerMask = RCCornerMaskAll ^ RCCornerMaskMinXMaxY;
        config.rect = _button.bounds;
        config.cornerRadius = 20;
        config.fillColor = UIColor.redColor.CGColor;
        [_button.layer setRoundCorner:config];
    }
    return _button;
}

- (UITextField *)textField {
    if (!_textField) {
        _textField = [[UITextField alloc] initWithFrame:CGRectMake(20, 155, 125, 125)];
        _textField.clearButtonMode = UITextFieldViewModeWhileEditing;
        _textField.placeholder = @"Input";
        RCConfiguration *config = [[RCConfiguration alloc] init];
        config.cornerMask = RCCornerMaskAll ^ RCCornerMaskMaxXMinY;
        config.rect = _textField.bounds;
        config.cornerRadius = 20;
        config.strokeColor = UIColor.redColor.CGColor;
        [_textField.layer setRoundCorner:config];
    }
    return _textField;
}

- (UIView *)emptyView {
    if (!_emptyView) {
        _emptyView = [[UIView alloc] initWithFrame:CGRectMake(155, 155, 125, 125)];
        RCConfiguration *config = [[RCConfiguration alloc] init];
        config.cornerMask = RCCornerMaskMaxXMinY | RCCornerMaskMinXMaxY | RCCornerMaskMaxXMaxY;
        config.rect = _emptyView.bounds;
        config.cornerRadius = 20;
        config.strokeColor = UIColor.redColor.CGColor;
        [_emptyView.layer setRoundCorner:config];
    }
    return _emptyView;
}

- (void)showRoundSectionTableViewController {
    RoundSectionTableViewController *vc = [[RoundSectionTableViewController alloc] init];
    [self presentViewController:vc animated:YES completion:nil];
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [self.view endEditing:YES];
}

@end

File: RoundCorner-Swift/RoundCorner-Swift/ViewController.swift
//
//  ViewController.swift
//  RoundCorner-Swift
//
//  Created by virusbee on 2021/4/25.
//

import UIKit

class ViewController: UIViewController {

    lazy var contentView: UIView = {
        let view = UIView(frame: CGRect(x: 0, y: 0, width: 300, height: 300))
        view.center = self.view.center
        
        var config = Configuration(rect: view.bounds)
        config.cornerMask = .All
        config.cornerRadius = 20
        config.strokeColor = UIColor.red.cgColor
        view.layer.setRoundCorner(config: config)
        
        return view
    }()
    
    lazy var label: UILabel = {
        let label = UILabel(frame: CGRect(x: 20, y: 20, width: 125, height: 125))
        var config = Configuration(rect: label.bounds)
        
        config.cornerMask = [.MinXMinY, .MaxXMinY, .MinXMaxY]
        config.cornerRadius = 20
        config.fillColor = UIColor.red.cgColor
        label.layer.setRoundCorner(config: config)
        
        return label
    }()
    
    lazy var button: UIButton = {
        let button = UIButton(frame: CGRect(x: 155, y: 20, width: 125, height: 125))
        button.setTitle("Touch Me", for: .normal)
        button.addTarget(self, action: #selector(showRoundSectionTableViewController), for: .touchUpInside)
        
        var config = Configuration(rect: button.bounds)
        config.cornerMask = [.MinXMinY, .MaxXMinY, .MaxXMaxY]
        config.cornerRadius = 20
        config.fillColor = UIColor.red.cgColor
        button.layer.setRoundCorner(config: config)
        
        return button
    }()
    
    lazy var textField: UITextField = {
        let textField = UITextField(frame: CGRect(x: 20, y: 155, width: 125, height: 125))
        textField.clearButtonMode = .whileEditing
        textField.placeholder = "Input"
        
        var config = Configuration(rect: textField.bounds)
        config.cornerMask = [.MinXMinY, .MaxXMaxY, .MinXMaxY]
        config.cornerRadius = 20
        config.strokeColor = UIColor.red.cgColor
        
        textField.layer.setRoundCorner(config: config)
        
        return textField
    }()
    
    lazy var emptyView: UIView = {
        let view = UIView(frame: CGRect(x: 155, y: 155, width: 125, height: 125))
        
        var config = Configuration(rect: view.bounds)
        config.cornerMask = [.MaxXMinY, .MinXMaxY, .MaxXMaxY]
        config.cornerRadius = 20
        config.strokeColor = UIColor.red.cgColor
        view.layer.setRoundCorner(config: config)
        
        return view
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.view.addSubview(contentView)
        contentView.addSubview(label)
        contentView.addSubview(button)
        contentView.addSubview(textField)
        contentView.addSubview(emptyView)
    }

    @objc func showRoundSectionTableViewController() {
        self.present(RoundSectionTableViewController(), animated: true, completion: nil)
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        self.view.endEditing(true)
    }
}

